rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+advertised by the cluster or 1 (or more) version greater than the version
+advertised by the cluster, an error will be shown and will require the user to
+use the `--skip-version-check` flag.
+
+```
+$ tsh login --proxy=proxy.example.com
+ERROR: Client tools are N major versions out of date, update to vX.Y.Z.
+
+Some functionality will not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+Use the "--skip-version-check" flag to bypass this check and attempt to connect
+to this cluster.
+```
+
+#### Self-managed client tools updates
+
+Cluster administrators that want to self-manage client tools updates will be
+able to watch for changes to client tools versions which can then be used to
+trigger other integrations (using MDM software like Jamf) to update the
+installed version of client tools on endpoints.
+
+```
+$ tctl autoupdate watch
+{"tools_version": "1.0.0"}
+{"tools_version": "1.0.1"}
+{"tools_version": "2.0.0"}
+
+[...]
+```
+
+Cluster administrators who prefer to use the Teleport API can create a watcher
+on the `types.ClusterMaintenanceConfig` resource and trigger other integrations
+on changes to the `tools_version` field.
+
+```
+// Create a watcher to monitor for changes to client tools version.
+watch, err := p.TeleportClient.NewWatcher(ctx, types.Watch{
+    Kinds: []types.WatchKind{
+        types.WatchKind{Kind: types.ClusterMaintenanceConfig},
+    },
+})
+if err != nil {
+    return trace.Wrap(err)
+}
+defer watch.Close()
+
+// Loop forever watching for client tools version changes.
+for {
+    select {
+    case e := <-watch.Events():
+        if e.Resource == nil {
+            continue
+        }
+        resource, ok := e.Resource.(types.ClusterMaintenanceConfig)
+        if !ok {
+            continue
+        }
+
+        // Process the version change by notifying downstream tooling (like
+        // Jamf).
+        if err := handle(resource); err != nil {
+            log.Printf("Failed to handle version change: %v", err)
+        }
+    case <-watch.Done():
+        fmt.Println("The watcher job is finished.")
+        return nil
+    }
+}
+```
+
+##### Cluster configuration
+
+> [!NOTE]
+> This resource will not be configurable on Teleport Cloud. It will be managed
+> by the Teleport Cloud team.
+
+The `types.ClusterMaintenanceConfig` resource will be updated to include a
+`tools_version` field. This field can be updated directly from the resource or
+using `tctl`.
+
+```
+kind: cluster_maintenance_config
+spec:
+  # tools_version is the version of client tools the proxy will advertise. Can
+  # be auto (match the version of the proxy), off (don't advertise any version),
+  # or an exact semver formatted version.
+  tools_version: auto | off | X.Y.Z
+
+  [...]
+```
```

Comment:
In Cloud context, do we want to block users from setting `tools_version` (e.g. reject anything different than `tools_version: auto`)?

Reply:
Yes, this resource will not be configurable on Cloud. We already do this for cluster networking config, and we'll extend this to `cluster_maintenance_config`.